### PR TITLE
add controller permissions to set blockOwnerDeletion

### LINKED ISSUE #51970: workload controllers can't create resources with GC admission on
owner ref admission checks permissions to decide if an ownerref is legal.  No controller has those permissions, so they can't create the resources they need to.

### PR DESCRIPTION
**Pull request by deads2k**
fixes https://github.com/kubernetes/kubernetes/issues/51970

`blockOwnerDeletion` requires delete permissions on the owner object.  This adds that permission for our controllers.

@kubernetes/sig-auth-misc 

```release-note
The OwnerReferencesPermissionEnforcement admission plugin now requires `update` permission on the `finalizers` subresource of the referenced owner in order to set `blockOwnerDeletion` on an owner reference.
```

### CODE REVIEW

**Comment by liggitt** [comment]
is there a related issue, or is this only with an admission plugin enabled?

**Comment by deads2k** [comment]
> is there a related issue, or is this only with an admission plugin enabled?

Only when its enabled.  We need to enable it for e2e tests, but I think I'd do that separately.

**Comment by liggitt** [comment]
blocking deletion of the owning object seems like it should require finalization permissions on the owning object more than deletion permissions

**Comment by liggitt** [comment]
this would be a great use case for a consistent `finalize` subresource on objects we could permission on, require GC-related operations to go through, etc.

**Comment by caesarxuchao** [comment]
> blocking deletion of the owning object seems like it should require finalization permissions on the owning object more than deletion permissions

sgtm.

@deads2k It looks like we haven't enabled the `OwnerReferencesPermissionEnforcement` admission plugin.Otherwise the missing permissions would had causing problems already. Shall we enable the plugin by default?

**Comment by liggitt** [comment]
I really don't want to give delete permissions to controllers that don't need them, simply so they can pass a synthetic permissions check for this admission plugin. Long-term, I'd like to see a consistent `finalizers` subresource that finalization operations go through. In the short-term, I'd rather see this do an authorization check of `update` permissions on a hypothetical `finalizers` subresource, and give that permission to the controllers.

**Comment by deads2k** [comment]
@liggitt updated to use update on `<resource>/finalizers`

File: plugin/pkg/admission/gc/gc_admission.go [hunk 1]
```diff
@@ -178,12 +178,12 @@ func (a *gcPermissionsEnforcement) ownerRefToDeleteAttributeRecords(ref metav1.O
...
 		ret = append(ret, authorizer.AttributesRecord{
 			User: attributes.GetUserInfo(),
-			Verb: "delete",
+			Verb: "update",
 			// ownerReference can only refer to an object in the same namespace, so attributes.GetNamespace() equals to the owner's namespace
 			Namespace:       attributes.GetNamespace(),
 			APIGroup:        groupVersion.Group,
 			APIVersion:      groupVersion.Version,
-			Resource:        mapping.Resource,
+			Resource:        mapping.Resource + "/finalizers",
```
Comment (liggitt) [inline plugin/pkg/admission/gc/gc_admission.go]: subresource field in this struct

Comment (deads2k) [reply plugin/pkg/admission/gc/gc_admission.go]: > subresource field in this struct

fixed.

File: plugin/pkg/admission/gc/gc_admission_test.go [hunk 2]
```diff
@@ -39,20 +39,29 @@ func (fakeAuthorizer) Authorize(a authorizer.Attributes) (bool, string, error) {
 		if a.GetVerb() == "delete" {
 			return false, "", nil
 		}
+		if a.GetVerb() == "update" && a.GetSubresource() == "/finalizers" {
```
Comment (liggitt) [inline plugin/pkg/admission/gc/gc_admission_test.go]: "/finalizers" or "finalizers"?

**Comment by liggitt** [comment]
this lgtm. @caesarxuchao, any final comments?

relatedly, can you open a doc PR for this admission plugin? https://github.com/kubernetes/kubernetes.github.io/issues/3588 noted we were missing documentation for it

**Comment by caesarxuchao** [comment]
The design and code lgtm. I'm concerned whether it's backward incompatible. I guess it breaks existing users that set blockOwnerDeletion, since no one has permission to update the finalizer subresources.

**Comment by liggitt** [comment]
Only if you enable this admission plugin, and if you did that, all your existing controllers were broken already.  Given that there was no documentation around this plug-in,  and of the default policy did not work with it, I doubt there are any existing users.  I think a release noted change is sufficient

**Comment by caesarxuchao** [comment]
Makes sense. LGTM.

**Comment by deads2k** [comment]
Tagging per comments

**Pull request merged**